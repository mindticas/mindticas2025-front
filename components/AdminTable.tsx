'use client';

import { Table, Box, Spinner } from '@chakra-ui/react';
import { ReactNode } from 'react';

interface ColumnDef<T> {
    key: string;
    header: string;
    render: (item: T) => ReactNode;
    align?: 'left' | 'center' | 'right';
    width?: string | number;
}

interface AdminTableProps<T> {
    data: T[];
    columns: ColumnDef<T>[];
    isLoading?: boolean;
    emptyMessage?: string;
    scrollable?: boolean;
}

export function AdminTable<T>({
    data,
    columns,
    isLoading = false,
    emptyMessage = 'No hay datos disponibles',
}: AdminTableProps<T>) {
    if (isLoading) {
        return (
            <Box
                display='flex'
                alignItems='center'
                justifyContent='center'
                height='300px'
            >
                <Spinner size='xl' />
            </Box>
        );
    }

    const TableContent = (
        <Table.Root size='sm' variant='outline' showColumnBorder>
            <Table.Header>
                <Table.Row bg='#F3F4F6'>
                    {columns.map((column) => (
                        <Table.ColumnHeader
                            key={column.key}
                            textAlign={column.align || 'center'}
                            width={column.width}
                            py={2}
                            fontWeight='bold'
                        >
                            {column.header}
                        </Table.ColumnHeader>
                    ))}
                </Table.Row>
            </Table.Header>
            <Table.Body>
                {data.length === 0 ? (
                    <Table.Row>
                        <Table.Cell
                            p={3}
                            colSpan={columns.length}
                            textAlign='center'
                        >
                            {emptyMessage}
                        </Table.Cell>
                    </Table.Row>
                ) : (
                    data.map((item, index) => (
                        <Table.Row
                            className='tr-table'
                            key={index}
                            borderBottomWidth={1}
                        >
                            {columns.map((column) => (
                                <Table.Cell
                                    key={`${column.key}-${index}`}
                                    textAlign={column.align || 'center'}
                                    px={2}
                                    width='fit-content'
                                >
                                    {column.render(item)}
                                </Table.Cell>
                            ))}
                        </Table.Row>
                    ))
                )}
            </Table.Body>
        </Table.Root>
    );

    return (
        <Table.ScrollArea borderWidth='1px' borderRadius='lg'>
            {TableContent}
        </Table.ScrollArea>
    );
}
