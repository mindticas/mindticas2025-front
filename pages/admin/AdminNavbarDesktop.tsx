import {
    Box,
    Flex,
    Link as ChakraLink,
    Button,
    Text,
    Image,
} from '@chakra-ui/react';
import { LogOut, Scissors } from 'lucide-react';
import React from 'react';
import NextLink from 'next/link';
import { AdminNavbarDesktopProps } from '@/interfaces/navItems/navItems';
import { handleRefresh } from '@/services/RefreshToken';
import CalendarImg from '@/public/google-calendar.png';
import { handleLogout } from '@/services/authService';

export default function DesktopViewAdmin({
    navItems,
    pathname,
}: AdminNavbarDesktopProps) {
    return (
        <Box bg='gray.800' color='white' shadow='lg'>
            <Flex
                maxW='7xl'
                mx='auto'
                px={{ base: 4, md: 6, lg: 8 }}
                h='16'
                align='center'
                justify='space-between'
            >
                {/* Logo and links */}
                <Flex align='center'>
                    <ChakraLink
                        display='flex'
                        alignItems='center'
                        _hover={{ textDecoration: 'none' }}
                        color='white'
                    >
                        <Scissors size={32} />
                        <Text ml='2' fontSize='xl' fontWeight='bold'>
                            Elegangster Admin
                        </Text>
                    </ChakraLink>
                    <Flex ml='10' align='center' gap='4'>
                        {navItems &&
                            navItems.map((item) => {
                                const Icon = item.icon;
                                const active = pathname === item.href;
                                const hasName = !!item.name;
                                return (
                                    <ChakraLink
                                        key={item.name}
                                        as={NextLink}
                                        px='3'
                                        py='2'
                                        href={item.href}
                                        rounded='md'
                                        fontSize='md'
                                        fontWeight='medium'
                                        bg={active ? 'gray.900' : 'transparent'}
                                        _hover={{
                                            bg: 'gray.700',
                                            textDecoration: 'none',
                                        }}
                                        display='flex'
                                        alignItems='center'
                                        color='white'
                                    >
                                        <Icon
                                            width={24}
                                            height={24}
                                            style={{
                                                marginRight: hasName
                                                    ? '8px'
                                                    : '0px',
                                            }}
                                        />
                                        {item.name}
                                    </ChakraLink>
                                );
                            })}
                    </Flex>
                </Flex>
                <Flex align='center' gap='4'>
                    {/* Logout button section for desktop */}
                    <Button
                        py='2'
                        px='3'
                        color='white'
                        bg='red.600'
                        _hover={{ bg: 'red.700' }}
                        fontWeight='bold'
                        display='flex'
                        alignItems='center'
                        gap='2'
                        onClick={handleLogout}
                    >
                        <LogOut size={32} />
                        Salir
                    </Button>
                    {/* Google calendar button section for desktop */}
                    <Image
                        boxSize='40px'
                        rounded='md'
                        cursor='pointer'
                        onClick={handleRefresh}
                        src={CalendarImg.src}
                        alt='Imagen de google calendar'
                    />
                </Flex>
            </Flex>
        </Box>
    );
}
