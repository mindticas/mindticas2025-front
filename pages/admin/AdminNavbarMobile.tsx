import { AdminNavbarMobileProps } from '@/interfaces/navItems/navItems';
import {
    Box,
    Flex,
    Link as ChakraLink,
    IconButton,
    Stack,
    Button,
    Text,
} from '@chakra-ui/react';
import { LogOut, Menu, Scissors } from 'lucide-react';
import NextLink from 'next/link';

export default function AdminNavbarMobile({
    navItems,
    pathname,
    isMobileMenuOpen,
    onToggleMenu,
}: AdminNavbarMobileProps) {
    return (
        <Box bg='gray.800' color='white' shadow='lg'>
            <Flex
                maxW='7xl'
                mx='auto'
                px={{ base: 4, md: 6, lg: 8 }}
                h='16'
                align='center'
                justify='space-between'
            >
                {/* Logo */}
                <ChakraLink
                    as={NextLink}
                    href='/admin/citas'
                    display='flex'
                    alignItems='center'
                    _hover={{ textDecoration: 'none' }}
                    color='white'
                >
                    <Scissors size={32} />
                    <Text ml='2' fontSize='xl' fontWeight='bold'>
                        Elegangster Admin
                    </Text>
                </ChakraLink>

                <IconButton
                    aria-label='Abrir menÃº'
                    variant='ghost'
                    size='2xl'
                    onClick={onToggleMenu}
                >
                    <Menu color='white' />
                </IconButton>
            </Flex>
            {/* Mobile menu */}
            {isMobileMenuOpen && (
                <Box pb='4' display={{ lg: 'none' }}>
                    <Stack px='2'>
                        {navItems.map((item) => {
                            const Icon = item.icon;
                            const active = pathname === item.href;
                            return (
                                <ChakraLink
                                    key={item.name}
                                    as={NextLink}
                                    href={item.href}
                                    px='3'
                                    py='2'
                                    rounded='md'
                                    fontSize='base'
                                    fontWeight='medium'
                                    bg={active ? 'gray.900' : 'transparent'}
                                    _hover={{
                                        bg: 'gray.700',
                                        textDecoration: 'none',
                                    }}
                                    display='flex'
                                    alignItems='center'
                                    onClick={onToggleMenu}
                                    color='white'
                                >
                                    <Icon
                                        width={20}
                                        height={20}
                                        style={{ marginRight: '8px' }}
                                    />
                                    {item.name}
                                </ChakraLink>
                            );
                        })}
                        <Button
                            variant='ghost'
                            justifyContent='flex-start'
                            px='3'
                            py='2'
                            fontSize='base'
                            fontWeight='medium'
                            color='gray.300'
                            _hover={{ bg: 'gray.700', color: 'white' }}
                        >
                            <LogOut size={20} />
                            Salir
                        </Button>
                    </Stack>
                </Box>
            )}
        </Box>
    );
}
