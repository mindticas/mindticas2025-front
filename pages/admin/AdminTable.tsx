'use client';

import { Table, Box, Spinner } from '@chakra-ui/react';
import { ReactNode, useEffect, useState } from 'react';
import Pagination from './Pagination';
import adminTableMessages from '@/constants/Admin/adminMessages';

interface ColumnDef<T> {
    key: string;
    header: string;
    render: (item: T) => ReactNode;
    align?: 'left' | 'center' | 'right';
    width?: string | number;
}

interface AdminTableProps<T> {
    data: T[];
    columns: ColumnDef<T>[];
    isLoading?: boolean;
    emptyMessage?: string;
    scrollable?: boolean;
    itemsPerPage?: number;
}

/**
 * A generic table component for displaying data in an admin interface.
 *
 * @template T - The type of the data items to be displayed in the table.
 *
 * @param {Object} props - The properties for the AdminTable component.
 * @param {T[]} props.data - The array of data items to be displayed in the table.
 * @param {Array<{ key: string; header: string; align?: 'left' | 'center' | 'right'; width?: string; render: (item: T) => React.ReactNode }>} props.columns -
 * An array of column definitions, where each column specifies a key, header text, optional alignment, optional width, and a render function for the cell content.
 * @param {boolean} [props.isLoading=false] - A flag indicating whether the table is in a loading state. If true, a loading spinner is displayed.
 * @param {string} [props.emptyMessage='No hay datos disponibles'] - The message to display when there is no data available.
 *
 * @returns {JSX.Element} The rendered table component, including a loading spinner, table content, or an empty message as appropriate.
 */
export default function AdminTable<T>({
    data = [],
    columns,
    isLoading = false,
    emptyMessage = adminTableMessages.emptyData.es,
    itemsPerPage = 10,
}: AdminTableProps<T>) {
    const [currentPage, setCurrentPage] = useState(1);
    const totalItems = data?.length;
    const totalPages = Math.ceil(totalItems / itemsPerPage);
    const startIndex = (currentPage - 1) * itemsPerPage;
    const endIndex = Math.min(startIndex + itemsPerPage, totalItems);
    const paginatedData = data.slice(startIndex, endIndex);

    // Reset the current page to 1 when the data changes
    useEffect(() => {
        if (currentPage !== 1) setCurrentPage(1);
    }, [data?.length]);

    // Adjust the current page if it exceeds the total number of pages
    useEffect(() => {
        if (currentPage > totalPages && totalPages > 0) {
            setCurrentPage(1);
        }
    }, [totalPages]);

    if (isLoading) {
        return (
            <Box
                display='flex'
                alignItems='center'
                justifyContent='center'
                height='300px'
            >
                <Spinner size='xl' />
            </Box>
        );
    }

    const TableContent = (
        <Table.Root size='sm' variant='outline' showColumnBorder>
            <Table.Header>
                <Table.Row bg='#F3F4F6'>
                    {columns &&
                        columns.map((column) => (
                            <Table.ColumnHeader
                                key={column.key}
                                textAlign={column.align || 'center'}
                                width={column.width}
                                py={2}
                                fontWeight='bold'
                            >
                                {column.header}
                            </Table.ColumnHeader>
                        ))}
                </Table.Row>
            </Table.Header>
            <Table.Body>
                {paginatedData && paginatedData?.length === 0 ? (
                    <Table.Row>
                        <Table.Cell
                            p={3}
                            colSpan={columns?.length}
                            textAlign='center'
                        >
                            {emptyMessage}
                        </Table.Cell>
                    </Table.Row>
                ) : (
                    paginatedData &&
                    paginatedData.map((item, index) => (
                        <Table.Row
                            className='tr-table'
                            key={index}
                            borderBottomWidth={1}
                        >
                            {columns &&
                                columns.map((column) => (
                                    <Table.Cell
                                        key={`${column.key}-${index}`}
                                        textAlign={column.align || 'center'}
                                        px={2}
                                        width='fit-content'
                                    >
                                        {column.render(item)}
                                    </Table.Cell>
                                ))}
                        </Table.Row>
                    ))
                )}
            </Table.Body>
        </Table.Root>
    );

    return (
        <Box>
            <Table.ScrollArea borderWidth='1px' borderRadius='lg'>
                {TableContent}
            </Table.ScrollArea>
            <Pagination
                currentPage={currentPage}
                totalItems={totalItems}
                itemsPerPage={itemsPerPage}
                onPageChange={setCurrentPage}
            />
        </Box>
    );
}
