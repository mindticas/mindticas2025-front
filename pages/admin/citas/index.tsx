import Layout from '@/components/Layout';
import React from 'react';

export default function Citas() {
    return (
        <div>
            <h1>Pagina de citas</h1>
        </div>
    );
}
