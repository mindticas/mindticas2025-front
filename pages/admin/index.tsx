export default function Admin() {
    return (
        <>
            <h1>Bienvenido al panel</h1>
        </>
    );
}
