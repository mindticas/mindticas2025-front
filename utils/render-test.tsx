import { Provider } from '@/components/ui/provider';
import { BookingProvider } from '@/context/BookingContext';
import { render as rtlRender } from '@testing-library/react';

export function render(ui: React.ReactNode) {
    return rtlRender(<>{ui}</>, {
        wrapper: (props: React.PropsWithChildren) => (
            <Provider>
                <BookingProvider>{props.children}</BookingProvider>
            </Provider>
        ),
    });
}
