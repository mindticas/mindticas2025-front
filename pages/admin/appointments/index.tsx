'use client';

import type React from 'react';
import { useState, useMemo, useEffect } from 'react';
import { Box, Text, useBreakpointValue } from '@chakra-ui/react';
import {
    getAppointmentById,
    getAppointments,
    updateAppointment,
} from '@/services/AppointmentService';
import { Appointment } from '@/interfaces/appointment/Appointment';
import { getTreatments } from '@/services/TreatmentService';
import { Treatment } from '@/interfaces/treatment/Treatment';
import { splitDateTimeFromISO } from '@/utils/dateUtils';
import { Toaster, toaster } from '@/components/ui/toaster';

import { SearchFilters } from '@/components/SearchFilters';
import { useFilters } from '@/hooks/useFilters';
import {
    AppointmentUpdate,
    FormDataType,
} from '@/interfaces/appointment/AppointmentUpdate';
import StatusBadge from './components/StatusBadge';
import AppointmentActions from './components/ActionButtons';
import AppointmentModal from './components/AppointmentModal';

import { DateTime } from 'luxon';
import AdminTable from '../AdminTable';

/**
 * The `CitasPage` component is the main page for managing appointments in the admin panel.
 * It provides functionality for viewing, filtering, editing, canceling, and completing appointments.
 * The page also includes a modal for creating or editing appointments and a table for displaying appointment data.
 *
 * @component
 *
 * @returns {JSX.Element} The rendered component for the appointments page.
 *
 */

export default function CitasPage() {
    const isSmallScreen = useBreakpointValue({ base: true, md: false });
    const [appointments, setAppointments] = useState<Appointment[]>([]);
    const { filters, handleFilterChange, handleStatusChange } = useFilters({
        name: '',
        treatments: '',
        date: '',
        status: '',
    });

    const [modalState, setModalState] = useState<{
        isOpen: boolean;
        mode: 'create' | 'edit' | 'cancel';
        appointment?: Appointment;
    }>({
        isOpen: false,
        mode: 'create',
    });

    const [isLoading, setIsLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);
    const [treatments, setTreatments] = useState<Treatment[]>([]);

    useEffect(() => {
        const fetchInitialData = async () => {
            try {
                const [appointmentsData, treatmentsData] = await Promise.all([
                    getAppointments(),
                    getTreatments(),
                ]);
                const adjustedAppointments = appointmentsData.map(
                    (appointment) => {
                        const adjustedScheduledStart = DateTime.fromISO(
                            appointment.scheduled_start,
                            { zone: 'utc' },
                        ).setZone('local');

                        return {
                            ...appointment,
                            scheduled_start: adjustedScheduledStart.toString(),
                        };
                    },
                );
                setAppointments(adjustedAppointments);
                setTreatments(treatmentsData);
            } catch (error) {
                setError(
                    'Error al cargar los datos. Inténtalo de nuevo más tarde.',
                );
            } finally {
                setIsLoading(false);
            }
        };

        fetchInitialData();
    }, []);

    const handleEdit = (id: number) => {
        const appointment = appointments.find((a) => a.id === id);

        if (appointment) {
            setModalState({
                isOpen: true,
                mode: 'edit',
                appointment,
            });
        }
    };

    const handleConfirm = async (data: {
        id?: number;
        name: string;
        treatment: string;
        date: string;
        time: string;
        tipAmount?: number;
        products?: number[];
    }) => {
        try {
            if (modalState.mode === 'edit' && data.id) {
                const originalAppointment = await getAppointmentById(data.id);
                const adjustedAppointment = {
                    ...originalAppointment,
                    scheduled_start: DateTime.fromISO(
                        originalAppointment.scheduled_start,
                        { zone: 'utc' },
                    )
                        .setZone('local')
                        .toFormat("yyyy-MM-dd'T'HH:mm"),
                };
                const updatedFields = getUpdatedFields(
                    data,
                    adjustedAppointment,
                );

                if (Object.keys(updatedFields).length == 0) {
                    toaster.create({
                        type: 'info',
                        duration: 3000,
                        title: 'No se detectaron cambios para actualizar',
                    });
                    return;
                }

                await updateAppointment(data.id, updatedFields);
                await refreshAppointmentState(data.id);

                toaster.create({
                    type: 'success',
                    duration: 5000,
                    title: 'Cita actualizada con éxito',
                });
            } else if (modalState.mode === 'create') {
                // Logic if you want to create a new appointment in the future for the moment not implemented
            }

            setModalState({ isOpen: false, mode: 'create' });
        } catch (error) {
            handleUpdateError(error);
        }
    };

    const getUpdatedFields = (
        formData: FormDataType,
        originalAppointment: Appointment,
    ) => {
        const updatedFields: Partial<AppointmentUpdate> = {};

        if (formData.name !== originalAppointment.customer?.name) {
            updatedFields.customer_name = formData.name;
        }

        const originalTreatmentIds = originalAppointment.treatments
            .map((t) => t.id)
            .sort()
            .join(',');
        const newTreatmentIds = formData.treatment
            .split(',')
            .map((id) => parseInt(id.trim()))
            .sort()
            .join(',');

        if (newTreatmentIds !== originalTreatmentIds) {
            updatedFields.treatments_id = formData.treatment
                .split(',')
                .map((id) => parseInt(id.trim()));
        }

        const newScheduledStart = `${formData.date}T${formData.time}`;

        if (newScheduledStart !== originalAppointment.scheduled_start) {
            const adjustedScheduledStart = DateTime.fromISO(newScheduledStart, {
                zone: 'local',
            })
                .setZone('utc')
                .toFormat("yyyy-MM-dd'T'HH:mm");
            updatedFields.scheduled_start = adjustedScheduledStart;
        }
        if (
            typeof formData.tipAmount !== 'undefined' &&
            Number(formData.tipAmount) !== Number(originalAppointment.tipAmount)
        ) {
            updatedFields.tipAmount = Number(formData.tipAmount);
        }

        if (
            Array.isArray(formData.products) &&
            JSON.stringify(formData.products.sort()) !==
                JSON.stringify(
                    (originalAppointment.products ?? [])
                        .map((p) => p.id)
                        .sort(),
                )
        ) {
            updatedFields.products = formData.products;
        }
        return updatedFields;
    };

    const refreshAppointmentState = async (appointmentId: number) => {
        const verifiedAppointment = await getAppointmentById(appointmentId);

        const adjustedAppointment = {
            ...verifiedAppointment,
            scheduled_start: DateTime.fromISO(
                verifiedAppointment.scheduled_start,
                { zone: 'utc' },
            )
                .setZone('local')
                .toFormat('yyyy/MM/dd, HH:mm'),
        };

        setAppointments((prev) =>
            prev.map((appointment) =>
                appointment.id === appointmentId
                    ? adjustedAppointment
                    : appointment,
            ),
        );
    };

    const handleUpdateError = (error: unknown) => {
        const errorMessage =
            error instanceof Error
                ? error.message
                : 'Error al actualizar la cita';
        toaster.create({
            type: 'error',
            duration: 5000,
            title: errorMessage,
        });
    };

    const filteredAppointments = useMemo(() => {
        return appointments.filter((appointment) => {
            return (
                appointment.customer?.name
                    .toLowerCase()
                    .includes(filters.name.toLowerCase()) &&
                appointment.treatments?.some((treatment) =>
                    treatment.name
                        .toLowerCase()
                        .includes(filters.treatments!.toLowerCase()),
                ) &&
                appointment.scheduled_start.includes(filters.date!) &&
                (filters.status === '' || appointment.status === filters.status)
            );
        });
    }, [appointments, filters]);

    const handleCancelClick = (id: number) => {
        const appointmentToCancel = appointments.find(
            (appointment) => appointment.id === id,
        );
        if (appointmentToCancel) {
            setModalState({
                isOpen: true,
                mode: 'cancel',
                appointment: appointmentToCancel,
            });
        }
    };

    const handleComplete = async (id: number) => {
        try {
            const completedAppointment = {
                status: 'completed',
            };
            await updateAppointment(id, completedAppointment);
            setAppointments((prev) =>
                prev.map((a) =>
                    a.id === id ? { ...a, status: 'completed' } : a,
                ),
            );

            toaster.create({
                type: 'success',
                duration: 5000,
                title: 'Cita completada con éxito',
            });
        } catch (error) {
            toaster.create({
                type: 'error',
                duration: 5000,
                title: 'Ocurrió un error al completar la cita, intenta de nuevo',
            });
        }
    };

    const handleCancelConfirm = async (id: number) => {
        try {
            const cancelAppointment = { status: 'canceled' };
            await updateAppointment(id, cancelAppointment);
            setAppointments((prev) =>
                prev.map((a) =>
                    a.id === id ? { ...a, status: 'canceled' } : a,
                ),
            );
            toaster.create({
                type: 'success',
                duration: 5000,
                title: 'Cita cancelada con éxito',
            });
        } catch (error) {
            toaster.create({
                type: 'error',
                duration: 5000,
                title: 'Error al cancelar la cita',
            });
        } finally {
            setModalState({ isOpen: false, mode: 'create' });
        }
    };

    const tableColumns = [
        {
            key: 'name',
            header: 'NOMBRE',
            render: (appointment: Appointment) =>
                appointment.customer?.name || 'Sin nombre',
            align: 'center' as const,
        },
        {
            key: 'treatment',
            header: 'TRATAMIENTO',
            render: (appointment: Appointment) =>
                appointment.treatments.map((t) => t.name).join(', ') ||
                'Sin tratamiento',
            align: 'center' as const,
        },
        {
            key: 'date',
            header: 'FECHA',
            render: (appointment: Appointment) =>
                new Date(appointment.scheduled_start).toLocaleDateString(
                    'es-ES',
                    {
                        day: '2-digit',
                        month: '2-digit',
                        year: 'numeric',
                    },
                ),
            align: 'center' as const,
        },
        {
            key: 'time',
            header: 'HORA',
            render: (appointment: Appointment) =>
                new Date(appointment.scheduled_start).toLocaleTimeString(
                    'es-ES',
                    {
                        hour: '2-digit',
                        minute: '2-digit',
                    },
                ),
            align: 'center' as const,
        },
        {
            key: 'status',
            header: 'ESTATUS',
            render: (appointment: Appointment) => (
                <StatusBadge
                    statusValue={appointment.status}
                    key={appointment.id}
                />
            ),
            align: 'center' as const,
        },
        {
            key: 'actions',
            header: 'ACCIONES',
            render: (appointment: Appointment) => (
                <AppointmentActions
                    appointmentId={appointment.id}
                    status={appointment.status}
                    onEdit={handleEdit}
                    onCancel={handleCancelClick}
                    onComplete={handleComplete}
                />
            ),
            align: 'center' as const,
        },
    ];

    return (
        <>
            <Box
                p={isSmallScreen ? 3 : 8}
                m={isSmallScreen ? 3 : 8}
                bg='white'
                borderRadius='lg'
                className='light'
            >
                <Text fontSize='2xl' fontWeight='bold' mb={4}>
                    Citas
                </Text>

                <div>
                    {error && (
                        <p
                            style={{
                                color: 'red',
                                textAlign: 'center',
                                padding: '10px',
                            }}
                        >
                            {error}
                        </p>
                    )}
                </div>

                <AppointmentModal
                    isOpen={modalState.isOpen}
                    onClose={() =>
                        setModalState({ isOpen: false, mode: 'create' })
                    }
                    mode={modalState.mode}
                    treatments={treatments}
                    initialData={
                        modalState.appointment
                            ? {
                                  id: modalState.appointment.id,
                                  name:
                                      modalState.appointment.customer?.name ||
                                      '',
                                  treatment: modalState.appointment.treatments
                                      .map((t) => t.id)
                                      .join(', '),
                                  date: splitDateTimeFromISO(
                                      modalState.appointment.scheduled_start,
                                  ).date,
                                  time: splitDateTimeFromISO(
                                      modalState.appointment.scheduled_start,
                                  ).time,
                                  tipAmmount:
                                      modalState.appointment.tipAmount ?? '',
                                  products:
                                      modalState.appointment.products?.map(
                                          (p) => p.id,
                                      ) ?? [],
                              }
                            : undefined
                    }
                    onConfirm={handleConfirm}
                    onCancel={handleCancelConfirm}
                />

                {/* Filters */}
                <SearchFilters
                    filters={filters}
                    onFilterChange={handleFilterChange}
                    onStatusChange={handleStatusChange}
                    showStatusFilter={true}
                    showDateFilter={true}
                    showTreatmentFilter={true}
                />

                <AdminTable
                    data={filteredAppointments}
                    columns={tableColumns}
                    isLoading={isLoading}
                />
            </Box>
            <Toaster />
        </>
    );
}
