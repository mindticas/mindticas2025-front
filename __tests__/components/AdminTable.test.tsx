import { AdminTable } from '@/pages/admin/AdminTable';
import { render } from '@/utils/render-test';

// Mock data for testing
const testData = [
    { id: 1, name: 'Item 1', value: 100 },
    { id: 2, name: 'Item 2', value: 200 },
];

const testColumns = [
    {
        key: 'name',
        header: 'Name',
        render: (item: any) => item.name,
        align: 'left' as const,
    },
    {
        key: 'value',
        header: 'Value',
        render: (item: any) => `$${item.value}`,
        align: 'right' as const,
    },
];

describe('AdminTable', () => {
    it('should render loading state correctly', () => {
        const { container } = render(
            <AdminTable data={[]} columns={testColumns} isLoading={true} />,
        );
        expect(container).toMatchSnapshot();
    });

    it('should render empty state correctly', () => {
        const { container } = render(
            <AdminTable
                data={[]}
                columns={testColumns}
                isLoading={false}
                emptyMessage='No items found'
            />,
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with data correctly', () => {
        const { container } = render(
            <AdminTable
                data={testData}
                columns={testColumns}
                isLoading={false}
            />,
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with custom empty message', () => {
        const { container } = render(
            <AdminTable
                data={[]}
                columns={testColumns}
                isLoading={false}
                emptyMessage='Custom empty message'
            />,
        );
        expect(container).toMatchSnapshot();
    });
});
