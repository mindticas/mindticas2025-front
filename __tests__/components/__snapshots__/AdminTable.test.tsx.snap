// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminTable should render empty state correctly 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-0"
  >
    <div
      class="css-7c1tm2"
    >
      <table
        class="chakra-table__root css-58tpe0"
      >
        <thead
          class="chakra-table__header css-hp6hec"
        >
          <tr
            class="chakra-table__row css-khv411"
          >
            <th
              class="chakra-table__columnHeader css-czbply"
            >
              Name
            </th>
            <th
              class="chakra-table__columnHeader css-1fdlycq"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="chakra-table__body css-0"
        >
          <tr
            class="chakra-table__row css-dj4s8p"
          >
            <td
              class="chakra-table__cell css-fb4w1i"
              colspan="2"
            >
              No items found
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`AdminTable should render loading state correctly 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-p6748q"
  >
    <span
      class="chakra-spinner css-1x6hz8j"
    />
  </div>
</div>
`;

exports[`AdminTable should render with custom empty message 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-0"
  >
    <div
      class="css-7c1tm2"
    >
      <table
        class="chakra-table__root css-58tpe0"
      >
        <thead
          class="chakra-table__header css-hp6hec"
        >
          <tr
            class="chakra-table__row css-khv411"
          >
            <th
              class="chakra-table__columnHeader css-czbply"
            >
              Name
            </th>
            <th
              class="chakra-table__columnHeader css-1fdlycq"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="chakra-table__body css-0"
        >
          <tr
            class="chakra-table__row css-dj4s8p"
          >
            <td
              class="chakra-table__cell css-fb4w1i"
              colspan="2"
            >
              Custom empty message
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`AdminTable should render with data correctly 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-0"
  >
    <div
      class="css-7c1tm2"
    >
      <table
        class="chakra-table__root css-58tpe0"
      >
        <thead
          class="chakra-table__header css-hp6hec"
        >
          <tr
            class="chakra-table__row css-khv411"
          >
            <th
              class="chakra-table__columnHeader css-czbply"
            >
              Name
            </th>
            <th
              class="chakra-table__columnHeader css-1fdlycq"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody
          class="chakra-table__body css-0"
        >
          <tr
            class="tr-table chakra-table__row css-1rw8fi1"
          >
            <td
              class="chakra-table__cell css-9tn4z"
            >
              Item 1
            </td>
            <td
              class="chakra-table__cell css-lwlazv"
            >
              $100
            </td>
          </tr>
          <tr
            class="tr-table chakra-table__row css-1rw8fi1"
          >
            <td
              class="chakra-table__cell css-9tn4z"
            >
              Item 2
            </td>
            <td
              class="chakra-table__cell css-lwlazv"
            >
              $200
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
